File: Intermediate/algorithms/QuickSort.py
#!/usr/bin/env python
# coding: utf-8

def partition(arr,low,high): 
    i = ( low-1 )        
    pivot = arr[high]     
  
    for j in range(low , high): 
  
        # If current element is smaller than or equal to pivot 
        if   arr[j] <= pivot: 
          
            # increment index of smaller element 
            i = i+1 
            arr[i],arr[j] = arr[j],arr[i] 
  
    arr[i+1],arr[high] = arr[high],arr[i+1] 
    return ( i+1 ) 
  


# arr[] --> Array to be sorted, 
# low  --> Starting index, 
# high  --> Ending index 
  
def quickSort(arr,low,high): 
    if low < high: 
  
        # pi is partitioning index, arr[p] is now at right place 
        pi = partition(arr,low,high) 
  
        # Separately sort elements before partition and after partition 
        quickSort(arr, low, pi-1) 
        quickSort(arr, pi+1, high) 
  

#Input format: 54 2 36 86 3
arr= [int(a) for a in input().split()]
n = len(arr) 
quickSort(arr,0,n-1) 
print ("Sorted array is:") 
for i in range(n): 
    print ("%d" %arr[i]) 

File: Intermediate/algorithms/quicksort.c
// C program to implement QuickSort
#include<stdio.h> 

void swap(int* a, int* b) 
{  
	*a = *a + *b; 
	*b = *a - *b;
	*a = *a - *b;
} 

int partition (int array[], int low, int high) 
{ 
	int pivot = array[high]; 
	int i = (low - 1);  

	for (int j = low; j <= high- 1; j++) 
	{ 
		if (array[j] <= pivot) 
		{ 
			i++;  
			swap(&arr[i], &arr[j]); 
		} 
	} 
	swap(&array[i + 1], &array[high]); 
	return (i + 1); 
} 

void quickSort(int array[], int low, int high) 
{ 
	if (low < high) 
	{ 
		
		int pi = partition(array, low, high); 

		quickSort(array, low, pi - 1); 
		quickSort(array, pi + 1, high); 
	} 
} 

void printArray(int array[], int size) 
{ 
	for (int i=0; i<size; i++) 
	printf("%d ",&array[i]); 
	printf(n); 
} 


int main() 
{ 
	int array[10],n=10;
	printf("Enter elements: /n ");
	for( int i=0;i<10;i++)
	scanf("%d",&array[i]);
	quickSort(array, 0, n-1); 
	printf("Sorted array: n"); 
printArray(array, n); 
	return 0; 
} 
